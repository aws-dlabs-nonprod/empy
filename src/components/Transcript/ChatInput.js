import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import Media from '@style/media';
import * as ActionTypes from '@constants/ActionTypes';
import { CSSTransition } from 'react-transition-group';
import { IconButton } from '@components/Button';
import { SoulMachinesContext } from '@contexts/SoulMachines';

const iconSend = require('@svg/icon-send.svg');

class ChatInput extends Component {

    static contextType = SoulMachinesContext;

    constructor() {
        super();

        this.state = {
            message: ''
        };

        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleChange = this.handleChange.bind(this);

        this.inputRef = React.createRef();
    }

    componentDidUpdate(prevProps) {
        const { isOpen } = this.props;

        // Focus the input field when we open
        if (prevProps.isOpen !== isOpen && isOpen) {
            this.inputRef.current.focus();
        }
    }

    handleSubmit(event) {
        if (typeof event === 'object') {
            event.preventDefault();
        }

        const { message } = this.state;

        if (message) {
            this.context.sendMessage(message);
        }

        this.setState({
            message: ''
        });
    }

    handleChange(e) {
        this.setState({
            message: e.target.value
        });
    }

    render() {
        const { message } = this.state;
        const { isOpen } = this.props;

        return (
            <CSSTransition in={ isOpen } timeout={ 3000 } unmountOnExit classNames="form">
                <StyledChatInput method="post" action="/" onSubmit={ this.handleSubmit }>
                    <input ref={ this.inputRef } onChange={ this.handleChange } value={ message } placeholder="Type a message or just speak it..." />
                    <IconButton icon={ iconSend } onClick={ this.handleSubmit } />
                </StyledChatInput>
            </CSSTransition>
        );
    }
}

const StyledChatInput = styled.form`
    bottom: 1.5rem;
    box-sizing: border-box;
    display: flex;
    flex-direction: row;
    height: 4.8rem;
    left: 1.5rem;
    overflow: hidden;
    position: fixed;
    right: 1.5rem;
    transition: width 0.3s ease-in-out;
    transition-property: opacity, width, transform;
    width: auto;
    will-change: opacity, width;
    z-index: 3;

    ${Media.desktop`
        left: auto;
        position: relative;
        right: auto;
        bottom: auto;
        width: 37rem;
    `}

    input {
        box-sizing: border-box;
        display: block;
        width: 100%;
        background: ${props => props.theme.colourForeground};
        border: 0.1rem solid ${props => props.theme.colourFaintBorder};
        border-right: none;
        border-radius: ${props => props.theme.radius} 0 0 ${props => props.theme.radius};
        height: 4.8rem;
        padding: 0 1.2rem;
        color: ${props => props.theme.textPrimary};
        font-size: 1.6rem;
        outline: none;
        
        &::placeholder {
            color: rgba(0, 0, 0, 0.4);
        }
    }

    ${IconButton} {
        background: ${props => props.theme.colourPrimary};
        border: ${props => `0.1rem solid ${props.theme.colourPrimary}`};
        border-radius: 0 ${props => props.theme.radius} ${props => props.theme.radius} 0;
        width: auto;
        height: 4.8rem;

        ${Media.desktop`
            &:hover {
                background: ${props => props.theme.colourPrimaryHover};
            }
        `}

        svg {
            fill: ${props => props.theme.colourForeground};
        }
    }

    /* Enter/Exit animation */
    &.form-enter {
        opacity: 0;
        transform: translateY(2rem);

        ${Media.desktop`
            transform: none;
            width: 0;
        `}
    }

    &.form-enter-done,
    &.form-enter-active {
        opacity: 1;
        transform: translateY(0);

        ${Media.desktop`
            transform: none;
            width: 37rem;
        `}
    }

    &.form-exit {
        opacity: 1;
        transform: translateY(0);

        ${Media.desktop`
            transform: none;
            width: 37rem;
        `}
    }

    &.form-exit-done,
    &.form-exit-active {
        opacity: 0;
        transform: translateY(2rem);

        ${Media.desktop`
            transform: none;
            width: 0;
        `}
    }
`;

function mapStateToProps(state) {
    return {
        isOpen: state.isTranscriptOpen
    };
}

function mapDispatchToProps(dispatch) {
    return {
        sendMessage: (message) => dispatch({
            type: ActionTypes.TRANSCRIPT_SEND_MESSAGE,
            message: message
        })
    };
}

const ConnectedChatInput = connect(
    mapStateToProps,
    mapDispatchToProps
)(ChatInput);

export {
    ConnectedChatInput as ChatInput,
    StyledChatInput
};
