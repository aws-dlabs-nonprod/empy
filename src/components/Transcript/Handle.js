import React from 'react';
import styled from 'styled-components';

const Handle = (props) => {
    const { className } = props;

    return (
        <div className={ className }>

        </div>
    );
};

export default styled(Handle)`
    position: absolute;
    height: 4.8rem;
    width: 100%;
    background: rgba(255, 0, 0, 0.1);
    top: -3.3rem;

    &:after {
        content: '';
        background: ${props => props.theme.colourBackground};
        border-radius: 1rem;
        height: 0.6rem;
        left: 0;
        margin: 0 auto;
        position: absolute;
        right: 0;
        top: 2.1rem;
        width: 3.1rem;
    }
`;
