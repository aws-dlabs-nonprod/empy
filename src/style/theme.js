export default {
    textPrimary: '#2D373E',
    textSecondary: '#E7E6E6',
    textBody: '#1A1A1A',
    textDark: '#621A4B',

    colourPrimary: '#D5002B',
    colourPrimaryHover: '#e24d6b',
    colourPrimaryTransparent: 'rgba(0, 174, 236, 0.5)',

    colourForeground: '#FFFFFF',
    colourForegroundHover: '#F9F9F8',
    colourForegroundTransparent: 'rgba(255, 255, 255, 0.2)',

    colourInactive: '#7F7F7F',
    colourBackground: '#E5E5E5',
    colourDivider: '#D7D2CB',
    colourFaintBorder: '#8C9296',

    radius: '0.3rem',
    splitHeight: '50vh'
};
