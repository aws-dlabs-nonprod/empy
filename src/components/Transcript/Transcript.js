import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import Media from '@style/media';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import SimpleBar from 'simplebar-react';
import * as ActionTypes from '@constants/ActionTypes';
import SimpleBarCSS from '@style/simplebar';
import animateScrollTo from 'animated-scroll-to';

import TranscriptMessage from '@components/Transcript/TranscriptMessage';
import { ChatInput, StyledChatInput } from '@components/Transcript/ChatInput';
import { IconButton } from '@components/Button';
import { SoulMachinesContext } from '@contexts/SoulMachines';
import { Indicator, StyledIndicator } from '@components/Indicator';

const iconClose = require('@svg/icon-close.svg');
const iconMic = require('@svg/icon-mic.svg');
const iconMicMute = require('@svg/icon-mic-mute.svg');

class Transcript extends Component {

    static contextType = SoulMachinesContext;

    constructor() {
        super();

        this.scrollRef = null;
        this.transcriptRef = React.createRef();
        this.contentRef = React.createRef();

        this.handleScroll = this.handleScroll.bind(this);
        this.handlePanelChange = this.handlePanelChange.bind(this);
        this.handleScrollToBottom = this.handleScrollToBottom.bind(this);
        this.handleToggleTranscript = this.handleToggleTranscript.bind(this);
        this.handleToggleMute = this.handleToggleMute.bind(this);
    }

    bindScrollRef(instance) {
        if (instance) {
            this.scrollRef = instance;
        }
    }

    handleToggleMute() {
        this.props.toggleMute();
    }

    handleToggleTranscript() {
        const { isOpen } = this.props;

        this.props.toggleShrinkPersona(!isOpen);

        setTimeout(() => {
            this.props.toggleTranscript(!isOpen);
        }, 200);
    }

    handleScroll() {
        const { scrollTop, scrollHeight, offsetHeight } = this.scrollRef;
        const { offsetHeight: contentHeight } = this.contentRef.current;
        const { current: transcript } = this.transcriptRef;

        const hasOverflowTop = scrollTop > 0 && contentHeight > offsetHeight;
        const hasOverflowBottom = (scrollHeight - scrollTop) >= offsetHeight + 15;

        /*
         * This isn't ideal, but if we set a state for it then performance will
         * dive as we trigger a React render every scroll step
         */
        transcript.classList.toggle('overflow-top', hasOverflowTop);
        transcript.classList.toggle('overflow-bottom', hasOverflowBottom);
    }

    handlePanelChange(direction, messageIndex, panel) {
        const { infoPanels } = this.props;

        if (!this.scrollRef) {
            return;
        }

        let targetPanelIndex = panel.index + direction;

        targetPanelIndex = Math.max(targetPanelIndex, 0);
        targetPanelIndex = Math.min(targetPanelIndex, infoPanels.length - 1);

        if (panel.index === targetPanelIndex) {
            return;
        }

        this.scrollToPanel(targetPanelIndex);
    }

    handleScrollToBottom() {
        this.scrollToBottom();
    }

    scrollToPanel(panelIndex) {
        const { transcript } = this.props;

        for (let i = 0; i < transcript.length; i++) {
            const { panel } = transcript[i];

            if (!panel) {
                continue;
            }

            if (panel.index === panelIndex) {
                const messageNode = document.querySelector(`#message-${i}`);

                // Position it in the middle of the transcript
                const targetScrollPosition = messageNode.offsetTop
                    - (this.scrollRef.clientHeight / 2)
                    + (messageNode.clientHeight / 2);

                this.scrollRef.scrollTo({
                    left: 0,
                    top: targetScrollPosition - 20,
                    behavior: 'smooth'
                });

                break;
            }
        }
    }

    scrollToBottom() {
        if (this.scrollRef) {
            animateScrollTo(this.scrollRef.scrollHeight, { element: this.scrollRef });
        }
    }

    componentDidMount() {
        const { eventBus } = this.context;

        eventBus.on('scroll:bottom', this.handleScrollToBottom);
    }

    componentDidUpdate(prevProps) {
        const { isOpen } = this.props;
        const { eventBus } = this.context;

        // Transcript toggled
        if (prevProps.isOpen !== isOpen) {
            if (isOpen) {
                eventBus.emit('scroll:bottom');
            }
        }
    }

    render() {
        const { isConnected, isOpen, isMuted, personaState } = this.props;

        return (
            <CSSTransition in={ isOpen && isConnected } timeout={ 1000 } unmountOnExit classNames="transcript">
                <StyledTranscript ref={ this.transcriptRef } { ...this.props }>
                    <IconButton
                        isToggled={ true }
                        className="close"
                        icon={ iconClose }
                        onClick={ this.handleToggleTranscript }
                    />

                    <IconButton
                        className="mute"
                        icon={ iconMic }
                        toggleIcon={ iconMicMute }
                        tipPosition="center"
                        onClick={ this.handleToggleMute }
                        isToggled={ isMuted }
                    />

                    <Indicator personaState={ personaState } />

                    <SimpleBar onScroll={ this.handleScroll }
                        scrollableNodeProps={{ ref: (instance) => this.bindScrollRef(instance) }}
                        style={{ overflowX: 'hidden' }}>
                        { this.renderItems() }
                    </SimpleBar>

                    <ChatInput />
                </StyledTranscript>
            </CSSTransition>
        );
    }

    renderItems() {
        const { transcript, infoPanels, activePanelIndex } = this.props;

        return (
            <div ref={ this.contentRef }>
                <TransitionGroup className="transcript-content">
                    { transcript.map((message, index) =>
                        <CSSTransition key={ index } timeout={ 1000 } classNames="item">
                            <TranscriptMessage
                                key={ `message-${index}`}
                                id={ `message-${index}`}
                                { ...message }
                                activePanelIndex={ activePanelIndex }
                                totalPanelCount={ infoPanels.length }
                                onContentResize={ () => {
                                    this.scrollToBottom();
                                }}
                                onPanelChange={ (direction) => {
                                    this.handlePanelChange(direction, index, message.panel);
                                }}
                            />
                        </CSSTransition>
                    ) }
                </TransitionGroup>
            </div>
        );
    }
};

const StyledTranscript = styled.div`
    background: ${props => props.theme.colourBackground};
    border-top: 0.1rem solid ${props => props.theme.colourDivider};
    bottom: 0;
    height: ${props => props.theme.splitHeight};
    left: 0;
    opacity: 0;
    position: absolute;
    top: auto;
    transform: translateY(130%);
    transition: opacity 0.5s ease-in-out, transform 0.5s ease-in-out;
    width: 100%;
    z-index: 2;

    ${Media.desktop`
        transition-duration: 0.5s, 0s;
        background: transparent;
        border: none;
        bottom: 10rem;
        left: 3rem;
        top: 11rem;
        width: 43.3rem;
        height: auto;
    `}

    ${StyledChatInput} {
        opacity: 0;
        transform: translateY(2rem);
        transition: opacity 0.3s ease-in-out, transform 0.3s ease-in-out;

        ${Media.desktop`
            display: none;
        `}
    }

    ${StyledIndicator} {
        position: absolute;
        margin: 0 auto;
        left: 0;
        right: 0;
        top: -6.4rem;
        height: 4.8rem;

        ${Media.desktop`
            display: none;
        `}
    }

    /* Transcript close button */
    .close {
        position: absolute;
        left: 1.5rem;
        top: -6.4rem;
        width: 4.8rem;
        height: 4.8rem;
        padding: 1.3rem;
        margin: 0;

        ${Media.desktop`
            display: none;
        `}
    }
    
    /* Mute toggle button */
    .mute {
        position: absolute;
        right: 1.5rem;
        top: -6.4rem;
        width: 4.8rem;
        height: 4.8rem;
        padding: 1.3rem;
        margin: 0;

        ${Media.desktop`
            display: none;
        `}
    }

    &:after,
    &:before {
        background: ${props => props.theme.colourDivider};
        content: '';
        height: 0.1rem;
        left: 0;
        opacity: 0;
        pointer-events: none;
        position: fixed;
        right: 0;
        top: 0;
        transition: opacity 0.2s ease-in-out, height 0.2s ease-in-out;
        z-index: 2;
    }

    &:before {
        display: none;

        ${Media.desktop`
            display: block;
            top: 0;
            width: 43.3rem;
            left: 0;
        `}
    }
    
    &:after {
        bottom: 7.7rem;
        top: auto;

        ${Media.desktop`
            bottom: 0;
            left: 0;
            width: 43.3rem;
        `}
    }

    &.overflow-top:before {
        opacity: 1;
    }

    &.overflow-bottom:after {
        opacity: 1;
    }
    
    .scroller {
        width: calc(100% - 3rem);
        height: calc(100% - 6rem);
        overflow: auto;
        padding: 0 1.5rem;
        position: absolute;

        ${Media.desktop`
            width: 100%;
            height: calc(100% - 1rem);
            padding: 0;
        `}
    }

    .transcript-content {
        padding: 1.5rem 0 0 1.5rem;

        ${Media.desktop`
            padding: 3rem 2rem 0 0;
        `}
    }
    
    /* Enter/Exit animation */
    &.transcript-enter {
        opacity: 0;
        transform: translateY(130%);

        ${StyledChatInput} {
            opacity: 0;
            transform: translateY(2rem);
        }

        ${Media.desktop`
            transform: translateY(0);
        `}
    }

    &.transcript-enter-done,
    &.transcript-enter-active {
        opacity: 1;
        transform: translateY(0);

        ${StyledChatInput} {
            opacity: 1;
            transform: translateY(0);
            transition-delay: 0.5s;
        }

        ${Media.desktop`
            transform: translateY(0);
        `}
    }

    &.transcript-exit {
        opacity: 1;
        transform: translateY(0);

        ${StyledChatInput} {
            opacity: 1;
            transform: translateY(0);
        }

        ${Media.desktop`
            transform: translateY(0);
        `}
    }

    &.transcript-exit-done,
    &.transcript-exit-active {
        opacity: 0;
        transform: translateY(130%);

        ${StyledChatInput} {
            opacity: 0;
            transform: translateY(2rem);
        }

        ${Media.desktop`
            transform: translateY(0);
        `}
    }

    ${SimpleBarCSS};
`;

function mapStateToProps(state) {
    return {
        isOpen: state.isTranscriptOpen,
        isMuted: state.isMuted,
        isConnected: state.isConnected,
        transcript: state.transcript,
        infoPanels: state.infoPanels,
        activePanelIndex: state.activePanelIndex,
        personaState: state.personaState
    };
}

function mapDispatchToProps(dispatch) {
    return {
        changePanel: (newIndex) => dispatch({
            type: ActionTypes.CHANGE_PANEL,
            newIndex
        }),

        toggleTranscript: (isToggled) => dispatch({
            type: ActionTypes.TRANSCRIPT_TOGGLE,
            isToggled
        }),

        toggleShrinkPersona: (isToggled) => dispatch({
            type: ActionTypes.TOGGLE_PERSONA_SHRINK,
            isToggled
        }),

        toggleMute: () => dispatch({
            type: ActionTypes.TOGGLE_MUTE
        })
    };
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Transcript);
