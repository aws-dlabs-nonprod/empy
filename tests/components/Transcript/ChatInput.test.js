import React from 'react';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import 'jest-styled-components';
import theme from '@style/theme';
import { ThemeProvider } from 'styled-components';

import { ChatInput } from '@components/Transcript/ChatInput';

const mockStore = configureStore([]);

describe('ChatInput', () => {

    test('renders correctly', () => {
        const store = mockStore({});

        const component = mount(
            <Provider store={ store }>
                <ThemeProvider theme={ theme }>
                    <ChatInput />
                </ThemeProvider>
            </Provider>
        );

        expect(component.find(ChatInput)).toMatchSnapshot();
    });

    test('handles transcript', () => {
        const store = mockStore({
            isTranscriptOpen: true
        });

        const component = mount(
            <Provider store={ store }>
                <ThemeProvider theme={ theme }>
                    <ChatInput />
                </ThemeProvider>
            </Provider>
        );

        expect(component.find(ChatInput)).toMatchSnapshot();
    });

});
