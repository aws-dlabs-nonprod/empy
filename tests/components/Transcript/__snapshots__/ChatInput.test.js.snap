// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatInput handles transcript 1`] = `
.c2 {
  background: #D5002B;
  border-radius: 0.3rem;
  box-sizing: border-box;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.6rem;
  height: 4.8rem;
  border: 0.1rem solid #D5002B;
  margin: 0 0 2rem 0;
  outline: none;
  padding: 0 1.8rem;
  position: relative;
  width: 100%;
  z-index: 3;
  -webkit-transition: background 0.2s ease,color 0.2s ease;
  transition: background 0.2s ease,color 0.2s ease;
  background: #F9F9F8;
  border-color: #D7D2CB;
  padding: 1.4rem 1.3rem;
}

.c2 svg {
  fill: #D5002B;
  height: 2rem;
  position: relative;
  width: 2rem;
}

.c0 {
  bottom: 1.5rem;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 4.8rem;
  left: 1.5rem;
  overflow: hidden;
  position: fixed;
  right: 1.5rem;
  -webkit-transition: width 0.3s ease-in-out;
  transition: width 0.3s ease-in-out;
  -webkit-transition-property: opacity,width,-webkit-transform;
  -webkit-transition-property: opacity,width,transform;
  transition-property: opacity,width,transform;
  width: auto;
  will-change: opacity,width;
  z-index: 3;
}

.c0 input {
  box-sizing: border-box;
  display: block;
  width: 100%;
  background: #FFFFFF;
  border: 0.1rem solid #8C9296;
  border-right: none;
  border-radius: 0.3rem 0 0 0.3rem;
  height: 4.8rem;
  padding: 0 1.2rem;
  color: #2D373E;
  font-size: 1.6rem;
  outline: none;
}

.c0 input::-webkit-input-placeholder {
  color: rgba(0,0,0,0.4);
}

.c0 input::-moz-placeholder {
  color: rgba(0,0,0,0.4);
}

.c0 input:-ms-input-placeholder {
  color: rgba(0,0,0,0.4);
}

.c0 input::placeholder {
  color: rgba(0,0,0,0.4);
}

.c0 .c1 {
  background: #D5002B;
  border: 0.1rem solid #D5002B;
  border-radius: 0 0.3rem 0.3rem 0;
  width: auto;
  height: 4.8rem;
}

.c0.form-enter {
  opacity: 0;
  -webkit-transform: translateY(2rem);
  -ms-transform: translateY(2rem);
  transform: translateY(2rem);
}

.c0.form-enter-done,
.c0.form-enter-active {
  opacity: 1;
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

.c0.form-exit {
  opacity: 1;
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

.c0.form-exit-done,
.c0.form-exit-active {
  opacity: 0;
  -webkit-transform: translateY(2rem);
  -ms-transform: translateY(2rem);
  transform: translateY(2rem);
}

@media (min-width:992px) and (min-height:0px) {
  .c2 {
    font-size: 1.8rem;
    padding: 0 2rem;
    width: 100%;
  }

  .c2:hover {
    background: #e24d6b;
  }

  .c2:hover svg {
    fill: #FFFFFF;
  }

  .c2:hover .ButtonTip__StyledButtonTip-zuccjc-0 {
    opacity: 0;
    -webkit-transform: translateY(0);
    -ms-transform: translateY(0);
    transform: translateY(0);
  }
}

@media (min-width:992px) and (min-height:0px) {
  .c2 {
    padding: 1.4rem 1.3rem;
  }

  .c2:hover {
    background: #FFFFFF;
  }

  .c2:hover svg {
    fill: #D5002B;
  }
}

@media (min-width:992px) and (min-height:0px) {
  .c0 {
    left: auto;
    position: relative;
    right: auto;
    bottom: auto;
    width: 37rem;
  }
}

@media (min-width:992px) and (min-height:0px) {
  .c0 .c1:hover {
    background: #e24d6b;
  }
}

@media (min-width:992px) and (min-height:0px) {
  .c0.form-enter {
    -webkit-transform: none;
    -ms-transform: none;
    transform: none;
    width: 0;
  }
}

@media (min-width:992px) and (min-height:0px) {
  .c0.form-enter-done,
  .c0.form-enter-active {
    -webkit-transform: none;
    -ms-transform: none;
    transform: none;
    width: 37rem;
  }
}

@media (min-width:992px) and (min-height:0px) {
  .c0.form-exit {
    -webkit-transform: none;
    -ms-transform: none;
    transform: none;
    width: 37rem;
  }
}

@media (min-width:992px) and (min-height:0px) {
  .c0.form-exit-done,
  .c0.form-exit-active {
    -webkit-transform: none;
    -ms-transform: none;
    transform: none;
    width: 0;
  }
}

<Connect(ChatInput)>
  <ChatInput
    isOpen={true}
    sendMessage={[Function]}
  >
    <CSSTransition
      classNames="form"
      in={true}
      timeout={3000}
      unmountOnExit={true}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={3000}
        unmountOnExit={true}
      >
        <ChatInput__StyledChatInput
          action="/"
          method="post"
          onSubmit={[Function]}
        >
          <StyledComponent
            action="/"
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "ChatInput__StyledChatInput-sc-12bvwwl-0",
                  "isStatic": false,
                  "lastClassName": "c0",
                  "rules": Array [
                    "bottom:1.5rem;box-sizing:border-box;display:flex;flex-direction:row;height:4.8rem;left:1.5rem;overflow:hidden;position:fixed;right:1.5rem;transition:width 0.3s ease-in-out;transition-property:opacity,width,transform;width:auto;will-change:opacity,width;z-index:3;",
                    "@media (min-width:",
                    "992",
                    "px) and (min-height:",
                    "0",
                    "px){",
                    "
        left: auto;
        position: relative;
        right: auto;
        bottom: auto;
        width: 37rem;
    ",
                    "}",
                    " input{box-sizing:border-box;display:block;width:100%;background:",
                    [Function],
                    ";border:0.1rem solid ",
                    [Function],
                    ";border-right:none;border-radius:",
                    [Function],
                    " 0 0 ",
                    [Function],
                    ";height:4.8rem;padding:0 1.2rem;color:",
                    [Function],
                    ";font-size:1.6rem;outline:none;&::placeholder{color:rgba(0,0,0,0.4);}}",
                    ".c1",
                    "{background:",
                    [Function],
                    ";border:",
                    [Function],
                    ";border-radius:0 ",
                    [Function],
                    " ",
                    [Function],
                    " 0;width:auto;height:4.8rem;",
                    "@media (min-width:",
                    "992",
                    "px) and (min-height:",
                    "0",
                    "px){",
                    "
            &:hover {
                background: ",
                    [Function],
                    ";
            }
        ",
                    "}",
                    "}&.form-enter{opacity:0;transform:translateY(2rem);",
                    "@media (min-width:",
                    "992",
                    "px) and (min-height:",
                    "0",
                    "px){",
                    "
            transform: none;
            width: 0;
        ",
                    "}",
                    "}&.form-enter-done,&.form-enter-active{opacity:1;transform:translateY(0);",
                    "@media (min-width:",
                    "992",
                    "px) and (min-height:",
                    "0",
                    "px){",
                    "
            transform: none;
            width: 37rem;
        ",
                    "}",
                    "}&.form-exit{opacity:1;transform:translateY(0);",
                    "@media (min-width:",
                    "992",
                    "px) and (min-height:",
                    "0",
                    "px){",
                    "
            transform: none;
            width: 37rem;
        ",
                    "}",
                    "}&.form-exit-done,&.form-exit-active{opacity:0;transform:translateY(2rem);",
                    "@media (min-width:",
                    "992",
                    "px) and (min-height:",
                    "0",
                    "px){",
                    "
            transform: none;
            width: 0;
        ",
                    "}",
                    "}",
                  ],
                },
                "displayName": "ChatInput__StyledChatInput",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "ChatInput__StyledChatInput-sc-12bvwwl-0",
                "target": "form",
                "toString": [Function],
                "usesTheme": true,
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            method="post"
            onSubmit={[Function]}
          >
            <form
              action="/"
              className="c0"
              method="post"
              onSubmit={[Function]}
            >
              <input
                onChange={[Function]}
                placeholder="Type a message or just speak it..."
                value=""
              />
              <Button__IconButton
                icon="mock-file"
                onClick={[Function]}
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "c1",
                        "isStatic": false,
                        "lastClassName": "c2",
                        "rules": Array [
                          "background:",
                          [Function],
                          ";border-radius:",
                          [Function],
                          ";box-sizing:border-box;color:",
                          [Function],
                          ";cursor:pointer;display:flex;justify-content:center;align-items:center;font-size:1.6rem;height:4.8rem;border:",
                          [Function],
                          ";margin:0 0 2rem 0;outline:none;padding:0 1.8rem;position:relative;width:100%;z-index:3;transition:background 0.2s ease,color 0.2s ease;",
                          "@media (min-width:",
                          "992",
                          "px) and (min-height:",
                          "0",
                          "px){",
                          "
        font-size: 1.8rem;
        padding: 0 2rem;
        width: 100%;

        &:hover {
            background: ",
                          [Function],
                          ";
            
            svg {
                fill: ",
                          [Function],
                          ";
            }

            ",
                          ".ButtonTip__StyledButtonTip-zuccjc-0",
                          " {
                opacity: ",
                          [Function],
                          ";
                transform: translateY(0);
            }
        }
    ",
                          "}",
                          "background:",
                          [Function],
                          ";border-color:",
                          [Function],
                          ";padding:1.4rem 1.3rem;",
                          "@media (min-width:",
                          "992",
                          "px) and (min-height:",
                          "0",
                          "px){",
                          "
        padding: 1.4rem 1.3rem;

        &:hover {
            background: ",
                          [Function],
                          ";

            svg {
                fill: ",
                          [Function],
                          ";
            }
        }
    ",
                          "}",
                          ";svg{fill:",
                          [Function],
                          ";height:2rem;position:relative;width:2rem;}",
                        ],
                      },
                      "displayName": "Button__IconButton",
                      "foldedComponentIds": Array [
                        "Button__BaseButton-sc-1t6ofhe-0",
                      ],
                      "render": [Function],
                      "styledComponentId": "c1",
                      "target": [Function],
                      "toString": [Function],
                      "usesTheme": true,
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  icon="mock-file"
                  onClick={[Function]}
                >
                  <Button
                    className="c1 c2"
                    disabled={false}
                    icon="mock-file"
                    isToggled={false}
                    onClick={[Function]}
                    secondary={false}
                    tip={null}
                    tipPosition="left"
                    toggleTip={null}
                  >
                    <button
                      aria-label={null}
                      className="c1 c2"
                      disabled={false}
                      onClick={[Function]}
                    >
                      <ReactSVG
                        evalScripts="never"
                        fallback={null}
                        loading={null}
                        onInjected={[Function]}
                        renumerateIRIElements={true}
                        src="mock-file"
                        svgClassName={null}
                        svgStyle={Object {}}
                        wrapper="div"
                      >
                        <div />
                      </ReactSVG>
                    </button>
                  </Button>
                </StyledComponent>
              </Button__IconButton>
            </form>
          </StyledComponent>
        </ChatInput__StyledChatInput>
      </Transition>
    </CSSTransition>
  </ChatInput>
</Connect(ChatInput)>
`;

exports[`ChatInput renders correctly 1`] = `
<Connect(ChatInput)>
  <ChatInput
    sendMessage={[Function]}
  >
    <CSSTransition
      classNames="form"
      timeout={3000}
      unmountOnExit={true}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={3000}
        unmountOnExit={true}
      />
    </CSSTransition>
  </ChatInput>
</Connect(ChatInput)>
`;
